DM-48788: Support ruff format

timj opened the pull request.

timj committed:
Update python version to 3.12

timj committed:
Use ruff format if ruff configuration is found

timj committed:
Split black and ruff into separate jobs

timj committed:
Add support for flake8/ruff check

Also disable isort if ruff is doing it

timj committed:
Tweak

timj committed:
Count results from grep

ktlim commented on .github/workflows/formatting.yaml:
```diff
@@ -1,32 +1,142 @@
+          use_ruff_check="false"
+          use_isort="true"
+          if [ -e pyproject.toml ]; then
+            if [ $(grep -c tool.ruff.format pyproject.toml) -gt 0 ]; then
```
Maybe use either `-F` or `\.` to avoid regexp treatment of the `.` characters.

I guess `pyproject.toml` is the best thing to base this on.  Are the strings given here always guaranteed to be present, even if there is no configuration needed for their associated tools?

timj replied:
We are in charge of our pyproject.toml and if someone thinks they enabled ruff and the action doesn't run it, it's easy to fix. That's where all the ruff configuration lives. The tricky thing is it depends on people using `tool.ruff.lint` and not arranging it in a slightly different hierarchy in the file. I was looking at TOML parsing tools but `yq` says it can parse TOML (so I could explicit look for `.tool.ruff.lint` regardless of how the file is arranged) but I haven't got it to work yet. The isort check is the most problematic one since the reality is that you can enable isort with "I" in the configuration selector but it's easier to look for `tool.ruff.lint.isort` (and we mostly want to set that anyhow to declare `lsst` as first party).

We are looking for `tool.ruff` in sconsUtils to decide whether to run flake8 vs ruff check during build time.

If we aren't using the pyproject.toml what's the alternative? Require parameters for the reusable action that specify the black vs ruff format etc?

timj replied:
`ruff check` always requires configuration.
`ruff format` doesn't really need configuration to do something useful but it seems a small price to pay to force a configuration block in the file to declare intent.

timj replied:
Looks like `yq` can sort of work (not the old PyPI one) if you first ask yq to convert the file to YAML.

```bash
$ yq -oy pyproject.toml | yq '.tool.ruff.lint.isort'
known-first-party:
  - lsst
```

timj replied:
This code works and there is an action plugin for installing yq:

```bash
# Decide to run ruff check if there is a ruff format configuration.
use_ruff_format="false"
use_ruff_check="false"
use_isort="true"
if [ -e pyproject.toml ]; then
    tmpfile="/tmp/pyproject.yaml"
    yq -oy pyproject.toml > ${tmpfile}
    if [ $(yq '(.tool.ruff | has("format"))' ${tmpfile}) == "true" ]; then
        use_ruff_format="true"
    fi
    if [ $(yq '(.tool.ruff | has("lint"))' ${tmpfile}) == "true" ]; then
        use_ruff_check="true"
    fi
    # No need to run isort if ruff check is already doing it.
    if [ $(yq '(.tool.ruff.lint | has("isort"))' ${tmpfile}) == "true" ]; then
        use_isort="false"
        use_ruff_check="true"
    fi
fi
if [ "${use_ruff_format}" == "true" ]; then
  echo "Using ruff format"
  echo "use_ruff_format=true"
else
  echo "Using black for formatting"
  echo "use_ruff_format=false"
fi
if [ "${use_ruff_check}" == "true" ]; then
  echo "Using ruff check for linting"
  echo "use_ruff_check=true"
else
  echo "Using flake8 for linting"
  echo "use_ruff_check=false"
fi
if [ "${use_isort}" == "true" ]; then
  echo "Using isort tool"
  echo "use_isort=true"
else
  echo "Using ruff check for isort"
fi
```

@ktlim happy to switch to this if you are worried about robustness since this guarantees that the hierarchy is handled however it's written.

timj replied:
I've switched to `yq` and that makes the logic more robust, especially since it now explicitly checks that `I` is enabled for `ruff check` rather than trying to guess. The only downside is the 14 seconds installing the ubuntu yq package... I've tried it out in a few packages and it seems to work. The one confusion is that the black and isort checks appear on the check reports as "skipped" and maybe people will think that is odd.

timj replied:
I've added the ability to force version overrides of all the packages and pinned things a little to avoid surprises.

timj committed:
Switch to using yq for interrogating ruff configuration

This is much more reliable in that it doesn't depend on
a particular implementation of hierarchy. It also allows
for an explicit look up of whether ruff check has enabled
isort.

timj committed:
Pin package versions and allow overrides
